Attatch users' photo consents on event & make sure consent for event exists before registration

LudvigHz opened the pull request:
Previously, we checked photo_consent status for an event in the frontend
by comparing the users registered photo consents with the event
start_time.

This among other things, led to a bug where the user does not have any
photo_consents for an event if the event is in the next semester, and
the user cannot attend the event if `use_consent=True`. This fixes that
by attaching and creating photo consents for that specific event.

This PR adds `photo_consents` on the `UserEventSerializer`, and will create the relevant photo
consent objects if they don't exist. Note that we don't create the photo consents unless the user
can register for the event _and_ the event enforces photo consent usage. This is just to avoid
creating photo consents for events far in the future, as well as events that have already happened.

eikhr reviewed (approved):
LGTM 📸

AdrianAndersen commented on lego/apps/events/serializers/events.py:
```diff
@@ -235,6 +238,17 @@ def get_pending_registration(self, obj):
 
+    def get_photo_consents(self, obj):
+        request = self.context.get("request", None)
+        if not request or not request.user.is_authenticated:
+            return []
+
+        if not obj.use_consent or len(obj.get_possible_pools(request.user)) == 0:
+            return []
```
Why is this not a single if statement? IMO looks better to have everything in one.

LudvigHz replied:
I split it up both to make it cleaner and to separate logic.

So the first one checks if the user is defined and the second checks if the user should have consents for the event.

I can merge them but I just find short conditional statements alot easier to read and understand.

ivarnakken replied:
I think neither solution is very elegant. I guess a comment of what you're validating here (like you're stating in the PR) would help.

LudvigHz replied:
I thought about adding a comment here, but it's very well explained in the commit message.

ivarnakken replied:
I know, but I'd prefer to not have to git blame something to understand ...

LudvigHz replied:
I'll add the comment.

But I will say that avoiding `git blame` is a really bad idea. It's super useful to get the full context of a change, with the authors intention and comment on the entire change and not just a single block of code.

ivarnakken replied:
Yes, I think combining code as docs and a good commit message is the optimal solution :)

AdrianAndersen commented on lego/apps/users/models.py:
```diff
@@ -552,24 +552,33 @@ class PhotoConsent(BasisModel):
-            semester=current_semester,
-            domain=constants.SOCIAL_MEDIA_DOMAIN,
-        )
+        consent_time = now
+        if time is not None:
+            consent_time = time
```
```suggestion
        consent_time = time if time is not None else now
```

AdrianAndersen reviewed (approved):
LGTM 👍🏻 Just a bit of code style.

ivarnakken reviewed (approved):
Makes sense :shipit: 

As always, please add some tests ❤️ I would also appreciate if you statically typed more of your code.

LudvigHz commented on lego/apps/users/models.py:
```diff
@@ -552,24 +552,33 @@ class PhotoConsent(BasisModel):
         unique_together = ("semester", "year", "domain", "user")
 
-    def get_consents(self, user):
+    def get_consents(self, user, *, time=None):
```
The `*` param just means that all the parameters following are named, so its not possible to pass an unnamed argument for a named parameter. It just makes the code more safe, especially when refactoring.

I don't see why the name here matters really? Since its quite explanatory and is not necessarily connected to the serializer, but if you really want to I can change it.

ivarnakken replied:
Cool, that makes sense. And the name is quite explanatory, indeed, but I would just like it to be perfect ..

erlingfn commented on lego/apps/users/models.py:
```diff
@@ -552,24 +552,33 @@ class PhotoConsent(BasisModel):
+
+        # Don't create PhotoConsent objects for the past
+        if consent_time >= now:
+            PhotoConsent.objects.get_or_create(
```
Could it be an idea to loop over the domains in the constants in case we add more domains in the future?

LudvigHz replied:
Yes!
I thought about it, but wanted this PR to just cover the bug it's fixing.
I could fix up all of the consent code as well, but it will make the PR larger as well as make more time.


Premature optimization is a thing, and there is no point in refactoring code before it's necessary. I think we should leave this until we _actually_ add another domain, then we can refactor and fix this code. Otherwise you just get caught in an endless refactoring and code change loop.

To quite Martin Fowler from the book _refactoring_[^1]:
> The first time you do something, you just do it. The second time you do something similar, you wince at the duplication, but you do the duplicate thing anyway. The third time you do something similar, you refactor.

[^1]: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

erlingfn replied:
Okay, fair enough

LudvigHz commented:
> As always, please add some tests :heart: I would also appreciate if you statically typed more of your code.

I did want to, but I just winced at the thought of writing tests for this, especially considering how horrible our test framework is atm :grimacing:. (I can see how much work it is to add the tests though)

static typing I can do though, thanks for the reminder!

LudvigHz commented:
@ivarnakken I have added some tests for the model, as well as the event API to cover the code paths introduced here. They should cover most of the cases here. And added some types, although there wasn't really that much to type (we really need stronger type checking to be able to do this properly :/)

ivarnakken commented on lego/apps/users/tests/test_models.py:
```diff
@@ -426,5 +430,22 @@ def test_get_consents_with_prior_consents(self):
+        user_consents = PhotoConsent.get_consents(self.test_user)
+        self.assertEqual(len(user_consents), 2)
+
+        today = timezone.now()
```
Kind of unnecessary variable, especially considering it's only used to calculate a different value.

ivarnakken commented on lego/apps/users/tests/test_models.py:
```diff
@@ -426,5 +430,22 @@ def test_get_consents_with_prior_consents(self):
+        consents = PhotoConsent.get_consents(
+            self.test_user, time=today - timedelta(weeks=52)
+        )
+        self.assertEqual(len(consents), 0)
```
`assertEmpty` is much needed here, but it seems that it's not available through our current use of the unittest library?

LudvigHz replied:
Yes, that would be nice but doesn't exist so :shrug:

ivarnakken commented on lego/apps/events/tests/test_events_api.py:
```diff
@@ -2034,3 +2034,104 @@ def test_filter_out_old(self):
+            "User should have the same amount of consent objects",
+        )
+
+        self.assertEquals(
```
See comment below.

ivarnakken reviewed (approved):
Thanks for the thorough tests! 😍 Just two very small comments, which I'll leave up to you to decide what to do with 🤗
